### Import Packages


In [1]:
# Import packages
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
import re

from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import seaborn as sn
import matplotlib.pyplot as plt
from imblearn.over_sampling import RandomOverSampler
from sklearn.svm import SVC


### Anonymise and Clean the Dataset

The original dataset was obtained from Vital Energi's VitalView system.  This is a live system used to assess the performance building Management Systems (BMS) and the heating, ventilation, and air conditiong systems they control.  It contains customer specific contract and building names.  
The following code takes in the original dataset and anonymises customer specific names and identification.  Only the redacted dataset has been provided with this code.
As the focus of the exercise is to predict datapoint category based on its name within a BMS I've dropped samples with generic sensor names and any NaN values within the 'Datapoint' and 'Category' columns only.

In [2]:
# Import original dataset, anonymise customer names with generic contract_n, drop NaN and generic values and save as anonymised

df = pd.read_csv('VitalView_dataset_cleaned.csv')

# Remove Generic Classes from Dataset and drop NaNs from Key Columns (Generic classes occur when correct type is not known)
df.replace({'Generic Analogue Sensor','Generic Digital Sensor','Default Auto'}, np.nan, inplace=True)
columns_to_dropna = ['Datapoint', 'Category']
df.dropna(subset=columns_to_dropna, inplace=True)

# Get the unique classes in the Contract/Site column
contract_column = 'Contract/Site'
contract_names = df[contract_column].unique()

# Create a dictionary to map classes to an anonimised name - contract_1...
class_to_anonymised = {class_name: f'contract_{i}' for i, class_name in enumerate(contract_names)}
df[contract_column] = df[contract_column].replace(class_to_anonymised)
contract_names = df[contract_column].unique()

# Save as New Anonymised .csv file
df.to_csv('VitalView_dataset_anonymised.csv', index=False)


### Load and Visualise the Redacted Dataset

In [3]:
# Import anonymised dataset & visualise
df = pd.read_csv('VitalView_dataset_anonymised.csv')
df.head()


Contract/Site Plantroom/Area          Datapoint       System  \
0    contract_0         Barnes   GenAhu_ClnExtFan  Ventilation   
1    contract_0         Barnes    GenAhu_Htg_Coil  Ventilation   
2    contract_0         Barnes      GenAhu_SupFan  Ventilation   
3    contract_0         Barnes  HWS_Destrat_Pump1          DHW   
4    contract_0         Barnes  HWS_Destrat_Pump2          DHW   

                  Category  
0            Enable Status  
1  Controlled Demand Speed  
2            Enable Status  
3            Enable Status  
4            Enable Status

### Drop Unimportant Columns
As the focus of the exercise is to predict datapoint category based on its name within a BMS I've dropped all columns excluding the 'Datapoint' which contains the shorthand name that we want to use as the predictor and 'Category' which is the class we want to predict.

In [4]:
# Drop unwanted columns
columns_to_drop = ['Contract/Site', 'Plantroom/Area', 'System']
df.drop(columns=columns_to_drop, inplace=True)
df.head()


Datapoint                 Category
0   GenAhu_ClnExtFan            Enable Status
1    GenAhu_Htg_Coil  Controlled Demand Speed
2      GenAhu_SupFan            Enable Status
3  HWS_Destrat_Pump1            Enable Status
4  HWS_Destrat_Pump2            Enable Status

### Visualise the Balance of Samples Within Each Category
It is important to understand the balance between number of samples in each category within the dataset.  If one or another category has significantly more or less samples it can affect the accuracy of model predictions.

In [5]:
# Check balance of categories and oversample if required
pd.set_option('display.max.rows', None) # Ensure df output isn't truncated

category_counts = df['Category'].value_counts()
category_df = pd.DataFrame({'Category Name': category_counts.index, 'Category Count': category_counts.values})
category_df


Category Name  Category Count
0                                  Enable Status            1280
1                        Controlled Demand Speed             458
2                                  Digital Alarm             295
3                            Comfort Temperature             221
4                                System Override             190
5                   Comfort Temperature Setpoint             178
6    Ventilation Supply Air Temperature Setpoint             135
7                                          Meter             112
8             Ventilation Supply Air Temperature              86
9                            Filter Dirty Status              81
10                           Frost Enable Status              69
11                           Ambient Temperature              64
12                         Frost Device Setpoint              58
13   Ventilation Heating Coil Return Temperature              55
14                                 Misc Setpoint              54
15     Ventilation Heating Coil Flow Temperature              50
16              Controlled Demand Speed Setpoint              38
17                               DHW Temperature              37
18            Secondary Vt Heating Flow Setpoint              36
19         Secondary Vt Heating Flow Temperature              36
20                      Frost Device Temperature              35
21                          DHW Flow Temperature              32
22                        DHW Return Temperature              30
23      Ventilation Heating Off-Coil Temperature              30
24                          OAT Eco Off Setpoint              27
25              Primary Heating Flow Temperature              26
26            Primary Heating Return Temperature              25
27                              Comfort Humidity              23
28                   Boiler Flue Gas Temperature              22
29                            Solar Active Power              22
30                          Comfort CO2 Setpoint              19
31                             Comfort CO2 Level              19
32       Secondary Vt Heating Return Temperature              17
33                                  DHW Setpoint              16
34               Ventilation Supply Air Humidity              16
35                           Comfort Air Quality              13
36                  Primary Chw Flow Temperature              11
37                  Ventilation Preheat Setpoint              10
38                Primary Chw Return Temperature              10
39          Comfort Temperature Setback Setpoint              10
40               Ventilation Preheat Temperature              10
41               Primary Heating Return Setpoint               9
42              Secondary Chw Return Temperature               9
43                        Pool Space Temperature               9
44               Ventilation Supply Air Pressure               9
45                  Comfort Air Quality Setpoint               8
46                Secondary Chw Flow Temperature               8
47      Secondary C T Heating Return Temperature               8
48                 Primary Heating Flow Setpoint               8
49                        PHE Return Temperature               7
50                         Irrandiance Totaliser               6
51      Ventilation Supply Air Humidity Setpoint               6
52            Secondary VT Flow Setback Setpoint               6
53                   DHW Pasteurisation Setpoint               5
54                           Run Hours Totaliser               5
55                        Cold Water Temperature               5
56                           Pool Space Humidity               5
57          Secondary Vt Heating Return Setpoint               5
58                       Occupancy Sensor Status               5
59  Underfloor Heating Flow Temperature Setpoint               5
60               Pool Space Temperature Setpoint               4
61   

There is a large disparity between categories with the majority count 1,280 and minority count only 3.  This will need to be considered when assessing model performance.

### Create a Custom Tokeniser

As the 'Datapoint' column consists of shorthand text and acronyms it is not possible to use standard text representation methods.  Instead a custom tokeniser will be required to convert key text into a bag of words that can be used to train a model.


In [6]:
# Create custom tokenizer

def customTokeniser (working_df):
    
    datapoint_text = working_df['Datapoint'].str.lower().tolist() # Convert to Lower Case

    # Create Custom List of Common Acronyms to Split
    acronyms_to_split = ['ahu', 'ext', 'fan', 'hws', 'lthw', 'vt', 'pri', 'sec', 'destrat', 'pump', 'oat', 'sup', 'htg', 'coil',
                        'ctrl', 'pool', 'temp', 'vlv', 'flw', 'rtn', 'pir', 'chw', 'en', 'flow', 'flw', 'frost', 'frst', 'hum', 
                        'humidifier', 'sp', 'sup', 'rm', 'spc', 'space', 'flow', 'return', 'setback', 'aq', 'supply', 'extract', 
                        'chiller', 'boiler', 'blr', 'dirty', 'tef', 'eco', 'phe', 'phx', 'reheat', 'heat', 'cool', 'run', 'status',
                        'cal', 'calorifier', 'dx', 'room', 'cef', 'def', 'valve', 'mthw', 'dstrt', 'zone', 'ac', 'aricon', 'dhw', 
                        'mtr', 'meter', 'kwh', 'kw', 'm3', 'press', 'kg', 'import', 'export', 'generated', 'lv', 'gas', 'parasitic',
                        'utilised', 'rejected', 'steam', 'flt', 'fault', 'sts', 'filter', 'fltr', 'trip', 'hand', 'hi', 'high', 
                        'lo', 'low', 'fresh', 'frsh', 'air', 'pnl', 'spd', 'dmd', 'speed', 'demand', 'clean', 'ct', 'lwr', 'str', 
                        'upr', 'mid', 'mcw', 'lights', 'lux', 'ht', 'cut', 'off', 'ambient', 'fire', 'alarm', 'warning', 'undrflr',
                        'clg', 'reht', 'drty', 'hb', 'cmn', 'occ', 'nsb', 'setpoint', 'fume', 'ppm', 'rh', 'district', 'dsrtct',
                        'demand', 'pump', 'imm', 'open', 'on', 'off', 'enabled', 'disabled', 'water', 'electric', 'elec', 'gas',
                        'tz', 'fcu', 'vav', 'mxb', 'dmpr', 'damper', 'override', 'ovrd', 'hours', 'hrs', 'oride', 'fail', 'reset',
                        'lphw', 'hthw', 'hphw', 'pos', 'shunt', 'shnt', 'ok', 'prove', 'winter', 'summer', 'dp','displace', 'setp',
                        'intake', 'filter', 'fltr', 'sts', 'fa', 'speed', 'spd', 'hand', 'pir', 'bag', 'pnl', 'panel', 'not', 'auto',
                        'night', 'setback', 'stbk', 'tz', 'current', 'power', 'energy', 'pos', 'irradiance', 'healthy', 'wtr', 
                        'water', 'comfort']

    # Custom tokenizer to split words using the list of acronyms
    def custom_tokenizer(text):
        for acronym in acronyms_to_split:
            # Use regex to find the acronym in the text and split it into separate words
            text = re.sub(r'\b{}\b'.format(acronym), ' '.join(acronym), text)
        words = text.split()  # Split the text into separate words
        return words

    # Create the CountVectorizer object with the custom tokenizer
    vectorizer = CountVectorizer(tokenizer=custom_tokenizer)

    # Fit and transform the text data into a bag of words representation
    bag_of_words = vectorizer.fit_transform(datapoint_text)

    # Convert the BoW representation into a DataFrame
    bow_df = pd.DataFrame(bag_of_words.toarray(), columns=vectorizer.get_feature_names_out())
    
    return bow_df


### Train and Asses an Initial Model

To gauge a rouhg understanding of potential model performance on the unbalanced dataset I've used a Random Forest and implemented a simple grid search to optimise model hyperparameters.

In [7]:
# Train an initial random forest

def train_RandomForest (df):
    
    X = customTokeniser(df)
    y = df['Category']


    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=1)
    # NOTE: Stratify ensures at least 1 of each class per split
                
        
    # Define the hyperparameter grid to search over
    param_grid = {
        'n_estimators': [20, 40, 80, 100, 200],
        'max_depth': [None, 2, 5, 10, 20],
        'min_samples_split': [5, 10, 20, 30],
        'min_samples_leaf': [5, 10, 20, 30, 40, 50]
    }

    # Do grid search with cross-validation
    rf = RandomForestClassifier(random_state=1)
    cv_rf = GridSearchCV(rf, param_grid, cv=5, n_jobs=-1)
    cv_rf.fit(X_train, y_train)

    # Get best hyperparameters
    best_params = cv_rf.best_params_

    # Train random forest model on entire training set using best hyperparameters
    rf = RandomForestClassifier(**best_params, random_state=1)
    rf.fit(X_train, y_train)
    return rf, X_train, X_test, y_train, y_test, best_params


In [8]:
# Make predictions on the test set - original unblanced dataset

rf, X_train, X_test, y_train, y_test, best_params = train_RandomForest(df)
y_pred = rf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get the classification report as a DataFrame
class_report_dict = classification_report(y_test, y_pred, target_names=y_test.unique(), output_dict=True)
class_report_df = pd.DataFrame(class_report_dict).transpose()

# Add the sample counts per category as a new column
category_sample_counts = y_test.value_counts().to_dict()
class_report_df['Sample Count'] = class_report_df.index.map(category_sample_counts)


C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:676: UserWarning: The least populated class in y has only 2 members, which is less than n_splits=5.
  warnings.warn(


Accuracy: 0.3099273607748184


C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Oversample to Balance Dataset

The prediction performance is very poor for the first model.  This could either be due to the imbalanced dataset or perhaps the Random Forest not being a good fit for the problem.  To test, oversample to see if performance improves.  This could lead to overfitting of the minority categories.


In [9]:
# Dataset is very imbalanced, oversample minority categories and retrain

# Find the number of samples in the majority class
majority_count = df['Category'].value_counts().max()

# Create an empty DataFrame to store the oversampled data
oversampled_df = pd.DataFrame()

# Group the DataFrame by the 'Category' column
grouped = df.groupby('Category')

# Loop through each category and oversample to match the majority count
for category, group in grouped:
    if len(group) < majority_count:
        # Resample the current category to match the majority count
        oversampled_group = group.sample(n=majority_count, replace=True, random_state=42)
        oversampled_df = pd.concat([oversampled_df, oversampled_group])
    else:
        # If the category already has enough samples, keep it as it is
        oversampled_df = pd.concat([oversampled_df, group])


# Print to Confirm Categories Now Balanced
category_counts = oversampled_df['Category'].value_counts()
category_df = pd.DataFrame({'Category Name': category_counts.index, 'Category Count': category_counts.values})
category_df


Category Name  Category Count
0                            Ambient Temperature            1280
1        Pool Space Temperature Setback Setpoint            1280
2               Secondary Chw Return Temperature            1280
3                 Secondary Chw Flow Temperature            1280
4       Secondary C T Heating Return Temperature            1280
5         Secondary C T Heating Flow Temperature            1280
6                            Run Hours Totaliser            1280
7             Primary Heating Return Temperature            1280
8                Primary Heating Return Setpoint            1280
9               Primary Heating Flow Temperature            1280
10                 Primary Heating Flow Setpoint            1280
11                Primary Chw Return Temperature            1280
12                  Primary Chw Flow Temperature            1280
13                     Primary Chw Flow Setpoint            1280
14                        Pool Water Temperature            1280
15                           Pool Water Setpoint            1280
16                 Pool Water Return Temperature            1280
17            Secondary VT Flow Setback Setpoint            1280
18            Secondary Vt Heating Flow Setpoint            1280
19         Secondary Vt Heating Flow Temperature            1280
20                  Ventilation Preheat Setpoint            1280
21   Ventilation Supply Air Temperature Setpoint            1280
22            Ventilation Supply Air Temperature            1280
23               Ventilation Supply Air Pressure            1280
24      Ventilation Supply Air Humidity Setpoint            1280
25               Ventilation Supply Air Humidity            1280
26               Ventilation Preheat Temperature            1280
27      Ventilation Heating Off-Coil Temperature            1280
28          Secondary Vt Heating Return Setpoint            1280
29   Ventilation Heating Coil Return Temperature            1280
30     Ventilation Heating Coil Flow Temperature            1280
31  Underfloor Heating Flow Temperature Setpoint            1280
32                               System Override            1280
33                            Solar Active Power            1280
34       Secondary Vt Heating Return Temperature            1280
35               Pool Space Temperature Setpoint            1280
36                        Pool Space Temperature            1280
37                   Boiler Flue Gas Temperature            1280
38                  Pool Space Humidity Setpoint            1280
39                        DHW Return Temperature            1280
40                   DHW Pasteurisation Setpoint            1280
41                          DHW Flow Temperature            1280
42              Controlled Demand Speed Setpoint            1280
43                       Controlled Demand Speed            1280
44                  Comfort Temperature Setpoint            1280
45          Comfort Temperature Setback Setpoint            1280
46                           Comfort Temperature            1280
47                     Comfort Humidity Setpoint            1280
48                              Comfort Humidity            1280
49                          Comfort CO2 Setpoint            1280
50                             Comfort CO2 Level            1280
51                  Comfort Air Quality Setpoint            1280
52                           Comfort Air Quality            1280
53                        Cold Water Temperature            1280
54                                  DHW Setpoint            1280
55                               DHW Temperature            1280
56                                 Digital Alarm            1280
57                                         Meter            1280
58          Pool Space Humidity Setback Setpoint            1280
59                           Pool Space Humidity            1280
60                        PHE Return Temperature            1280
61   

In [10]:
# Retrain model with oversampled dataset

X = customTokeniser(oversampled_df)
y = oversampled_df['Category']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=1)

# Initialize the Random Forest Classifier
rf_classifier = RandomForestClassifier(n_estimators=100, max_depth=50, max_features='sqrt', random_state=1)

# Train the Random Forest Classifier
rf_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = rf_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.6254815924657534


The model accuracy appears much better, however this could be related to overfitting on the oversampled minority classes so visualise performance by category to check.

In [11]:
# Visualise model performance by category

class_report_dict = classification_report(y_test, y_pred, target_names=y_test.unique(), output_dict=True)
class_report_df = pd.DataFrame(class_report_dict).transpose()
class_report_df


C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\Richard.Buckley\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


precision    recall  f1-score  \
Secondary Chw Return Temperature               1.000000  0.383104  0.553977   
Controlled Demand Speed Setpoint               1.000000  0.234818  0.380328   
Controlled Demand Speed                        1.000000  1.000000  1.000000   
OAT Eco Off Setpoint                           1.000000  0.709559  0.830108   
DHW Return Temperature                         1.000000  1.000000  1.000000   
Underfloor Heating Flow Temperature Setpoint   1.000000  0.605578  0.754342   
Secondary Vt Heating Return Setpoint           1.000000  0.412229  0.583799   
Ventilation Preheat Temperature                1.000000  0.340040  0.507508   
Lux Level Setpoint                             1.000000  1.000000  1.000000   
Ambient Temperature                            0.000000  0.000000  0.000000   
PHE Return Temperature                         1.000000  1.000000  1.000000   
Comfort CO2 Setpoint                           0.000000  0.000000  0.000000   
Secondary Vt Heating Flow Setpoint             0.000000  0.000000  0.000000   
Ventilation Supply Air Temperature             1.000000  0.116190  0.208191   
Comfort CO2 Level                              1.000000  0.298311  0.459538   
Frost Enable Status                            0.855196  1.000000  0.921947   
Secondary Vt Heating Flow Temperature          1.000000  0.254302  0.405488   
Frost Device Setpoint                          1.000000  0.522073  0.686003   
Pool Space Humidity Setback Setpoint           1.000000  0.075290  0.140036   
Occupancy Sensor Status                        0.000000  0.000000  0.000000   
Boiler Flue Gas Temperature                    0.000000  0.000000  0.000000   
Comfort Humidity Setpoint                      0.000000  0.000000  0.000000   
Comfort Temperature Setpoint                   1.000000  0.037037  0.071429   
Primary Chw Flow Temperature                   1.000000  0.130435  0.230769   
Primary Heating Return Temperature             0.000000  0.000000  0.000000   
Secondary Vt Heating Return Temperature        1.000000  1.000000  1.000000   
Ventilation Preheat Setpoint                   1.000000  1.000000  1.000000   
Comfort Temperature                            0.978389  1.000000  0.989076   
Secondary Chw Flow Temperature                 0.000000  0.000000  0.000000   
Secondary C T Heating Return Temperature       1.000000  0.035928  0.069364   
System Override                                1.000000  0.490835  0.658470   
Pool Space Temperature                         1.000000  1.000000  1.000000   
Run Hours Totaliser                            1.000000  1.000000  1.000000   
Ventilation Heating Coil Flow Temperature      1.000000  1.000000  1.000000   
Ventilation Temperature Setback Setpoint       1.000000  1.000000  1.000000   
Pool Space Temperature Setpoint                1.000000  1.000000  1.000000   
Pool Water Setpoint                            1.000000  1.000000  1.000000   
Meter                                          1.000000  1.000000  1.000000   
Irrandiance Totaliser                          1.000000  1.000000  1.000000   
Pool Space Humidity Setpoint                   1.000000  1.000000  1.000000   
Pool Water Temperature                         1.000000  1.000000  1.000000   
Enable Status                                  1.000000  1.000000  1.000000   
DHW Pasteurisation Setpoint                    1.000000  1.000000  1.000000   
Comfort Humidity                               1.000000  1.000000  1.000000   
Frost Device Temperature                       1.000000  1.000000  1.000000   
Comfort Temperature Setback Setpoint           1.000000  1.000000  1.000000   
Misc Setpoint                                  1.000000  0.630739  0.773562   
Pool Water Return Temperature                  1.000000  1.000000  1.000000   
Primary Chw Return Temperature                 1.000000  0.722753  0.839068   
Comfort Air Quality Setpoint                   1.000000  1.000000  1.000000   
Ventilation Heati

### Retrain Model on Majority Classes Only
Model performance is significantly affected by overfitting on the minority categories, as a test retrain the model on the majority categories only to see if performance improves.

In [12]:
# Try predicting on majority categories only

# Drop cateogies with < 50 Sample
category_counts = df['Category'].value_counts()
categories_to_drop = category_counts[category_counts < 50].index
df_majority = df[~df['Category'].isin(categories_to_drop)]

# Retrain and evaluate model
rf, X_train, X_test, y_train, y_test, best_params = train_RandomForest(df_majority)
y_pred = rf.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.3775811209439528


### Create a Binary Vector Tokeniser
The performance of the Random Forest is still relatively poor.  Instead of oversampling the original dataset, or removing the minority categories, it may be more beneficial to build a dedicated training set by randomly selecting a fixed number of samples from each category.  The BoW representation won't be suitable for this as the training and test sets will need to be split prior to being tokenised.  This will result in a different number of vectors per dataset.  Instead I've created a binary vector tokeniser that will ensure the same number of vectors between the dedicated training dataset and the test dataset.


In [13]:
# Create new tokeniser to generate binary vectors based on whether each shorthand/acronym exists or not

def binary_vector_tokenizer (working_df):
    
    binary_vectors_list = []
    
    # Create Custom List of Common Acronyms to Split
    acronyms_to_split = ['ahu', 'ext', 'fan', 'hws', 'lthw', 'vt', 'pri', 'sec', 'destrat', 'pump', 'oat', 'sup', 'htg', 'coil',
                        'ctrl', 'pool', 'temp', 'vlv', 'flw', 'rtn', 'pir', 'chw', 'en', 'flow', 'flw', 'frost', 'frst', 'hum', 
                        'humidifier', 'sp', 'sup', 'rm', 'spc', 'space', 'flow', 'return', 'setback', 'aq', 'supply', 'extract', 
                        'chiller', 'boiler', 'blr', 'dirty', 'tef', 'eco', 'phe', 'phx', 'reheat', 'heat', 'cool', 'run', 'status',
                        'cal', 'calorifier', 'dx', 'room', 'cef', 'def', 'valve', 'mthw', 'dstrt', 'zone', 'ac', 'aricon', 'dhw', 
                        'mtr', 'meter', 'kwh', 'kw', 'm3', 'press', 'kg', 'import', 'export', 'generated', 'lv', 'gas', 'parasitic',
                        'utilised', 'rejected', 'steam', 'flt', 'fault', 'sts', 'filter', 'fltr', 'trip', 'hand', 'hi', 'high', 
                        'lo', 'low', 'fresh', 'frsh', 'air', 'pnl', 'spd', 'dmd', 'speed', 'demand', 'clean', 'ct', 'lwr', 'str', 
                        'upr', 'mid', 'mcw', 'lights', 'lux', 'ht', 'cut', 'off', 'ambient', 'fire', 'alarm', 'warning', 'undrflr',
                        'clg', 'reht', 'drty', 'hb', 'cmn', 'occ', 'nsb', 'setpoint', 'fume', 'ppm', 'rh', 'district', 'dsrtct',
                        'demand', 'pump', 'imm', 'open', 'on', 'off', 'enabled', 'disabled', 'water', 'electric', 'elec', 'gas',
                        'tz', 'fcu', 'vav', 'mxb', 'dmpr', 'damper', 'override', 'ovrd', 'hours', 'hrs', 'oride', 'fail', 'reset',
                        'lphw', 'hthw', 'hphw', 'pos', 'shunt', 'shnt', 'ok', 'prove', 'winter', 'summer', 'dp','displace', 'setp',
                        'intake', 'filter', 'fltr', 'sts', 'fa', 'speed', 'spd', 'hand', 'pir', 'bag', 'pnl', 'panel', 'not', 'auto',
                        'night', 'setback', 'stbk', 'tz', 'current', 'power', 'energy', 'pos', 'irradiance', 'healthy', 'wtr', 
                        'water', 'comfort']
    
    working_df['Datapoint'] = working_df['Datapoint'].str.lower() # Convert to Lower Case
    
    for sample in working_df['Datapoint']:
        binary_vectors = []
        
        for acronym in acronyms_to_split:
            is_present = int(acronym in sample)
            binary_vectors.append(is_present)
        
        binary_vectors_list.append(binary_vectors)
    
    return binary_vectors_list


# Get binary vectors for each sentence (acronym)
binary_vectors_list = binary_vector_tokenizer(df)

#print("Binary vectors for each sentence (acronym):")
#for i, vectors in enumerate(binary_vectors_list):
#    print(f"Sentence {i+1}: {vectors}")


In [14]:
# Retrain and evaluate using the binary vector tokeniser and a dedicated balanced dataset for model training

# Group by category
grouped = df.groupby('Category')

# Randomly sample 50 samples from each category
samples_per_category = 50
sampled_data = []

for category, group in grouped:
     if len(group) >= samples_per_category:
         sampled_group = group.sample(samples_per_category)
     else:
         # If a category has fewer than 50 samples, you can choose to either duplicate samples
         num_samples_to_duplicate = samples_per_category - len(group)
         duplicated_samples = group.sample(num_samples_to_duplicate, replace=True)
         sampled_group = pd.concat([group, duplicated_samples])
     sampled_data.append(sampled_group)

# Combine sampled data from all categories into df for training
balanced_training_df = pd.concat(sampled_data)

X_train = binary_vector_tokenizer(balanced_training_df)
y_train = balanced_training_df['Category']
X_test = binary_vector_tokenizer(df)
y_test = df['Category']

 
# Define the hyperparameter grid to search over
param_grid = {
     'n_estimators': [20, 40, 80, 100, 200],
     'max_depth': [None, 2, 5, 10, 20],
     'min_samples_split': [5, 10, 20, 30],
     'min_samples_leaf': [5, 10, 20, 30, 40, 50]
}

# Do grid search with cross-validation
rf = RandomForestClassifier(random_state=1)
cv_rf = GridSearchCV(rf, param_grid, cv=5, n_jobs=-1)
cv_rf.fit(X_train, y_train)

# Get best hyperparameters
best_params = cv_rf.best_params_

# Train random forest model on entire training set using best hyperparameters
rf = RandomForestClassifier(**best_params, random_state=1)
rf.fit(X_train, y_train)

# Make predictions on the test set
y_pred = rf.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get the classification report as a DataFrame
class_report_dict = classification_report(y_test, y_pred, target_names=y_test.unique(), output_dict=True)
class_report_df = pd.DataFrame(class_report_dict).transpose()

# Add the sample counts per category as a new column
category_sample_counts = y_test.value_counts().to_dict()
class_report_df['Sample Count'] = class_report_df.index.map(category_sample_counts)


Accuracy: 0.6493097602325019


In [15]:
# Visualise model performance by category

class_report_df = class_report_df.sort_values(by='Sample Count', ascending=False)
class_report_df


precision    recall  f1-score  \
Enable Status                                  0.936170  0.687500  0.792793   
Controlled Demand Speed                        0.488889  1.000000  0.656716   
Digital Alarm                                  1.000000  1.000000  1.000000   
Comfort Temperature                            0.653061  1.000000  0.790123   
System Override                                0.741379  0.767857  0.754386   
Comfort Temperature Setpoint                   0.131579  1.000000  0.232558   
Ventilation Supply Air Temperature Setpoint    0.243056  0.921053  0.384615   
Meter                                          0.666667  1.000000  0.800000   
Ventilation Supply Air Temperature             0.964384  0.768559  0.855407   
Filter Dirty Status                            0.350000  1.000000  0.518519   
Frost Enable Status                            0.375000  0.789474  0.508475   
Ambient Temperature                            0.444444  0.800000  0.571429   
Frost Device Setpoint                          0.333333  0.750000  0.461538   
Ventilation Heating Coil Return Temperature    0.406780  0.888889  0.558140   
Misc Setpoint                                  0.898876  0.987654  0.941176   
Ventilation Heating Coil Flow Temperature      1.000000  0.592593  0.744186   
Controlled Demand Speed Setpoint               0.546875  0.972222  0.700000   
DHW Temperature                                1.000000  1.000000  1.000000   
Secondary Vt Heating Flow Setpoint             0.250000  1.000000  0.400000   
Secondary Vt Heating Flow Temperature          0.934426  0.982759  0.957983   
Frost Device Temperature                       0.586207  0.739130  0.653846   
DHW Flow Temperature                           0.923077  0.750000  0.827586   
DHW Return Temperature                         0.739130  0.918919  0.819277   
Ventilation Heating Off-Coil Temperature       0.714286  1.000000  0.833333   
OAT Eco Off Setpoint                           0.796078  0.688136  0.738182   
Primary Heating Flow Temperature               0.444444  1.000000  0.615385   
Primary Heating Return Temperature             0.055556  1.000000  0.105263   
Comfort Humidity                               1.000000  0.866667  0.928571   
Boiler Flue Gas Temperature                    0.750000  1.000000  0.857143   
Solar Active Power                             0.990566  0.777778  0.871369   
Comfort CO2 Setpoint                           0.210526  0.800000  0.333333   
Comfort CO2 Level                              0.650000  0.764706  0.702703   
Secondary Vt Heating Return Temperature        0.809524  0.971429  0.883117   
DHW Setpoint                                   0.454545  1.000000  0.625000   
Ventilation Supply Air Humidity                0.250000  0.900000  0.391304   
Comfort Air Quality                            0.993377  0.789474  0.879765   
Primary Chw Flow Temperature                   0.600000  1.000000  0.750000   
Ventilation Preheat Setpoint                   0.320000  1.000000  0.484848   
Primary Chw Return Temperature                 1.000000  1.000000  1.000000   
Comfort Temperature Setback Setpoint           1.000000  1.000000  1.000000   
Ventilation Preheat Temperature                0.500000  0.956522  0.656716   
Primary Heating Return Setpoint                0.222222  1.000000  0.363636   
Secondary Chw Return Temperature               0.909091  1.000000  0.952381   
Pool Space Temperature                         0.770833  0.672727  0.718447   
Ventilation Supply Air Pressure                0.954545  0.840000  0.893617   
Comfort Air Quality Setpoint                   0.833333  1.000000  0.909091   
Secondary Chw Flow Temperature                 0.423077  1.000000  0.594595   
Secondary C T Heating Return Temperature       0.774194  0.923077  0.842105   
Primary Heating Flow Setpoint                  1.000000  1.000000  1.000000   
PHE Return Temperature                         0.975472  0.403906  0.571271   
Irrandiance Total

### Model with an SVM and Compare Performance
Using a binary vector tokeniser and dedicated balanced dataset to train the model does provide better results.  Whilst the overall acuracy is still relatively low at 66%, and there is still some overfitting on the smallest minority categories, the predictions across categories are generally more realistic.  The lower accuracy is likely now an aspect of using a Random Forest with sparse binary vectors, to check this I've created an SVM model below as these are generally better at handing sparse binary vectors.  Again, I've used a simple grid search to tune model hyperparameters.


In [16]:
# Train SVM model using the binary vector tokeniser and dedicated balanced dataset for model training

# Group by category
grouped = df.groupby('Category')

# Randomly sample 50 samples from each category
samples_per_category = 50
sampled_data = []

for category, group in grouped:
     if len(group) >= samples_per_category:
         sampled_group = group.sample(samples_per_category)
     else:
         # If a category has fewer than 20 samples, you can choose to either duplicate samples
         num_samples_to_duplicate = samples_per_category - len(group)
         duplicated_samples = group.sample(num_samples_to_duplicate, replace=True)
         sampled_group = pd.concat([group, duplicated_samples])
     sampled_data.append(sampled_group)

# Combine sampled data from all categories
balanced_df = pd.concat(sampled_data)

X_train = binary_vector_tokenizer(balanced_df)
y_train = balanced_df['Category']
X_test = binary_vector_tokenizer(df)
y_test = df['Category']

 
# Create SVM model
svm_model = SVC()

# Set up the parameter grid for grid search
param_grid = {
    'C': [0.1, 1, 10],                # Regularization parameter values to try
    'kernel': ['linear', 'rbf'],      # Kernel types to try
}
    
    
# Create the GridSearchCV object
grid_search = GridSearchCV(svm_model, param_grid, cv=3, scoring='accuracy')

# Perform grid search on the training data
grid_search.fit(X_train, y_train)

# Get the best SVM model with the optimal hyperparameters
best_svm_model = grid_search.best_estimator_

# Make predictions on the test data using the best model
y_pred = best_svm_model.predict(X_test)

# Evaluate the model
category_sample_counts = y_test.value_counts().to_dict()

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Get the classification report as a DataFrame
class_report_dict = classification_report(y_test, y_pred, target_names=y_test.unique(), output_dict=True)
class_report_df = pd.DataFrame(class_report_dict).transpose()

# Add the sample counts per category as a new column
category_sample_counts = y_test.value_counts().to_dict()
class_report_df['Sample Count'] = class_report_df.index.map(category_sample_counts)


Accuracy: 0.8936788568660693


In [17]:
# Visualise performance by category
class_report_df = class_report_df.sort_values(by='Sample Count', ascending=False)
class_report_df


precision    recall  f1-score  \
Enable Status                                  0.859155  0.953125  0.903704   
Controlled Demand Speed                        1.000000  1.000000  1.000000   
Digital Alarm                                  1.000000  1.000000  1.000000   
Comfort Temperature                            0.864865  1.000000  0.927536   
System Override                                0.849206  0.955357  0.899160   
Comfort Temperature Setpoint                   0.333333  1.000000  0.500000   
Ventilation Supply Air Temperature Setpoint    0.486486  0.947368  0.642857   
Meter                                          1.000000  1.000000  1.000000   
Ventilation Supply Air Temperature             0.960373  0.899563  0.928974   
Filter Dirty Status                            0.777778  1.000000  0.875000   
Frost Enable Status                            0.846154  0.578947  0.687500   
Ambient Temperature                            0.714286  1.000000  0.833333   
Frost Device Setpoint                          0.666667  1.000000  0.800000   
Ventilation Heating Coil Return Temperature    0.794118  1.000000  0.885246   
Misc Setpoint                                  0.952381  0.987654  0.969697   
Ventilation Heating Coil Flow Temperature      0.531915  0.925926  0.675676   
Controlled Demand Speed Setpoint               0.972973  1.000000  0.986301   
DHW Temperature                                1.000000  1.000000  1.000000   
Secondary Vt Heating Flow Setpoint             1.000000  1.000000  1.000000   
Secondary Vt Heating Flow Temperature          0.964286  0.931034  0.947368   
Frost Device Temperature                       0.656250  0.913043  0.763636   
DHW Flow Temperature                           1.000000  0.750000  0.857143   
DHW Return Temperature                         1.000000  0.945946  0.972222   
Ventilation Heating Off-Coil Temperature       1.000000  1.000000  1.000000   
OAT Eco Off Setpoint                           0.921769  0.918644  0.920204   
Primary Heating Flow Temperature               0.571429  1.000000  0.727273   
Primary Heating Return Temperature             0.461538  1.000000  0.631579   
Comfort Humidity                               0.967742  1.000000  0.983607   
Boiler Flue Gas Temperature                    1.000000  0.888889  0.941176   
Solar Active Power                             0.991803  0.896296  0.941634   
Comfort CO2 Setpoint                           1.000000  0.800000  0.888889   
Comfort CO2 Level                              0.800000  0.941176  0.864865   
Secondary Vt Heating Return Temperature        0.971429  0.971429  0.971429   
DHW Setpoint                                   1.000000  1.000000  1.000000   
Ventilation Supply Air Humidity                0.562500  0.900000  0.692308   
Comfort Air Quality                            0.983607  0.947368  0.965147   
Primary Chw Flow Temperature                   0.750000  1.000000  0.857143   
Ventilation Preheat Setpoint                   0.320000  1.000000  0.484848   
Primary Chw Return Temperature                 1.000000  1.000000  1.000000   
Comfort Temperature Setback Setpoint           1.000000  1.000000  1.000000   
Ventilation Preheat Temperature                0.825000  0.956522  0.885906   
Primary Heating Return Setpoint                0.666667  1.000000  0.800000   
Secondary Chw Return Temperature               1.000000  1.000000  1.000000   
Pool Space Temperature                         0.911111  0.745455  0.820000   
Ventilation Supply Air Pressure                0.800000  0.960000  0.872727   
Comfort Air Quality Setpoint                   1.000000  1.000000  1.000000   
Secondary Chw Flow Temperature                 0.846154  1.000000  0.916667   
Secondary C T Heating Return Temperature       0.862069  0.961538  0.909091   
Primary Heating Flow Setpoint                  1.000000  1.000000  1.000000   
PHE Return Temperature                         0.987091  0.896094  0.939394   
Irrandiance Total

### Summary and Recomendations
In summary, due to the significant imbalance between the different categories in the original dataset, training a model with simple train/text split or oversampling minority categories does not perform well.  Instead, a dedicated training set had to be created from a smaller yet representative sample from each category.  This combined with a custom dictionary and tokeniser to convert key text from the 'Datapoint' category and an SVM model achieved good results with overall accuracy of 90%.  However, there is still some evidence of overfitting on the smallest minority categories.
It is recommended that further data be exported from VitalView to provide a minimum of 100 samples for categories from which a more robust training set can be constructed.
